---
output: 
  github_document:
    toc: true
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# Phydon

<!-- badges: start -->

<!-- badges: end -->

Estimating maximum growth rates of bacteria from genomic data with **Phydon**.

# Introduction

Phydon is a R package that estimates maximum growth rates of bacteria/archaea from genomic data and phylogenetic information. This package implements a method detailed in [Our paper](https://www.biorxiv.org/content/10.1101/2024.10.03.616540v1), leveraging phylogenetic signals and the genomic statistical factor, [codon usage bias (CUB)](https://github.com/jlw-ecoevo/gRodon2) (see [gRodon](https://github.com/jlw-ecoevo/gRodon2) for details), to enhance the accuracy of growth rate estimations. By integrating these advanced techniques, Phydon provides a robust tool for microbial growth rate analysis based on genomic information. 

# EGPO database

We developed a comprehensive database of temperature-corrected maximum growth rates for 111,349 species-representative genomes from GTDB v220, generated using Phydon and available as [EGPO](https://github.com/xl0418/Phydon/releases/tag/v1.0.0). This represents the largest temperature corrected maximum growth rate prediction database to date.



# Installation

Phydon installation is available from GitHub with:

```{r, eval = FALSE}
if (!requireNamespace("devtools", quietly = TRUE))
  install.packages("devtools")
devtools::install_github("xl0418/Phydon")
```

> [!IMPORTANT]\
> Phydon has several essential dependencies, including `gRodon2`, `ape`, `Biostrings`, `picante`, etc. Please make sure you have installed these packages before using Phydon.

```{r, eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

if (!requireNamespace("devtools", quietly = TRUE))
  install.packages("devtools")

BiocManager::install("Biostrings")
BiocManager::install("coRdon")
install.packages("matrixStats") 

devtools::install_github("jlw-ecoevo/gRodon2")
```


# Minimal Examples

## Identified Genomes

When users' genomes are identified by accession numbers found in the GTDB, the `user_tree` argument is not required. Below is a minimal example illustrating how to use Phydon:

```{r setup1, warning=FALSE, eval=FALSE}
library(Phydon)
# load the example data
data_info <- read.csv(system.file("extdata", "known.csv", package = "Phydon"))

#### To get the directory of the genomic data  ####
#### This is only for internal data as directory of genomic data is ad hoc per user's system ####
#### For user's input data, see below for details ####
genomes <- data_info$genome_name
gene_location <- c()
for (i in 1:length(genomes)) {
  gene_location[i] <- system.file("extdata", paste0("known/",genomes[i],"/",genomes[i],".ffn"), package = "Phydon")
}


# Generate some genomes with user picked names.
# The genome sequences are provided.
# In this case, only gRodon estimation will be used.
data_info$gene_location <- gene_location
data_info <- rbind(data_info, data.frame(gene_location = data_info$gene_location, genome_name = c(1:11)))

### temperature (optional)
# data_info$temperature <- rep(20, nrow(data_info))

# estimate the maximum growth rate in two modes
result_full <- Phydon(data_info, gRodon_mode="full")
result_metagenome <- Phydon(data_info, gRodon_mode="metagenome")

```

## Unidentified Genomes

When users define their own genome names (i.e., the genome is not identified by an accession number in the GTDB), the `user_tree` argument is required. Below is a minimal example illustrating how to use Phydon in this case:

```{r setup2, warning=FALSE, eval=FALSE}
library(Phydon)
# load the example data
data_info <- read.csv(system.file("extdata", "unknown.csv", package = "Phydon"))
# load the example tree
user_tree <- ape::read.tree(system.file("extdata", "usertree.tree", package = "Phydon"))

#### To get the directory of the genomic data  ####
#### This is only for internal data as directory of genomic data is ad hoc per user's system ####
#### For user's input data, see below for details ####
genomes <- data_info$genome_name
gene_location <- c()
for (i in 1:length(genomes)) {
  gene_location[i] <- system.file("extdata", paste0("unknown/",genomes[i],"/",genomes[i],".ffn"), package = "Phydon")
}
data_info$gene_location <- gene_location


# estimate the maximum growth rate
result <- Phydon(data_info, user_tree)


```


# Directory Structure for Genomic Data

> [!IMPORTANT]\
> For both modes, gRodon requires annotated genomes. Users must annotate their genomes before using Phydon. We recommend using [Prokka](https://github.com/tseemann/prokka) for genome annotation. The directory structure for genomic data should be organized as follows:

-   `genefiles/`
    -   `genome1/`
        -   `genome1.ffn`
        -   `genome1.gff`
        -   `genome1_CDS_names.txt`
    -   `genome2/`
        -   `genome2.ffn`
        -   `genome2.gff`
        -   `genome2_CDS_names.txt`
    -   `genome3/`
        -   `genome3.ffn`
        -   `genome3.gff`
        -   `genome3_CDS_names.txt`
    -   ...

> [!NOTE]\
> - The `genome1.ffn` file is the FASTA file of the genome, while the `genome1.gff` file is the corresponding annotation file. These files can be obtained during the genome annotation process. 

> - The `genome1_CDS_names.txt` file contains the names of the coding sequences (CDSs) in the genome. Phydon will attempt to generate this file using the `sed` command. If this fails, users will need to create the file manually. `sed` is typically available on Linux and macOS systems. For Windows users, you can install `sed` from [here](https://sourceforge.net/projects/gnuwin32/files/sed/4.2.1/). 

> - For more details, please refer to the [gRodon2](https://github.com/jlw-ecoevo/gRodon2) package.

> We recommend running Phydon on Linux or macOS to avoid potential issues with generating the `genome1_CDS_names.txt` file.




# Usage

## Arguments

- `data_info_df`: A data frame containing detailed information about the genomes to be analyzed. This must include relevant metadata for each genome (e.g., accession numbers, species identifiers, etc.). See the section "Data Frame Format" below for specific requirements.


- `user_tree`: A phylogenetic tree in Newick format (`Phylo class` in R) that represents the evolutionary relationships between the user's genomes and species from the GTDB (Genome Taxonomy Database). If users provide the genome accession numbers, the function will attempt to automatically retrieve a corresponding tree. If no accession numbers are given, users must supply the tree manually. This can be generated by running GTDB-Tk on the user's genome data, with the resulting tree typically found in the /output/classify/ folder after processing. The default value is NULL.

- `tax`: The taxonomic domain of the species being analyzed. Acceptable values are "bacteria" or "archaea". This parameter determines which evolutionary models are applied. The default value is "bacteria".

- `regression_mode`: Specifies the regression model to be used for estimating growth rates. Options are: "arithmetic_mean": Averages the growth rate estimates using an arithmetic mean with a continuous P value (see ). "geometric_mean": Uses the geometric mean for growth rate estimation (recommended for skewed data). "arithmetic_mean_binaryP": Averages the growth rate estimates using an arithmetic mean with a binary P value. "geometric_mean_binaryP": Uses the geometric mean for growth rate estimation with a binary P value. The default is "geometric_mean".

- `gRodon_mode`: The mode of gRodon predictions for genome-based growth rate estimates. Options are: "full": For full genome data (i.e., complete genomes). "metagenome": For metagenomic data (i.e., incomplete or assembled genomes). The default is "full".

### `data_info_df`: The genomic data information

The input data requires a data frame with the following columns:

-   `gene_location`: The directory path to the annotated genomic data. Each genome's corresponding FASTA file should be named using its accession number. For example, the FASTA file for the genome with accession number `RS_GCF_000024245.1` should be named `RS_GCF_000024245.1.ffn`.

-   `genome_name`: Either the accession number of the genome for which the user wants to predict the growth rate, or a user-defined name. If an accession number is provided, it is expected to exist in the GTDB database. If the accession number is not in GTDB, or if a user-defined name is used, Phydon will treat the genome as a unique species outside of GTDB. In such cases, **ONLY** gRodon-based predictions will be generated, while all other predictions will return `NA`. Refer to the example code for proper usage.

-   `temperature`: (**Optional**) This column is used in gRodon predictions. If temperature data is not provided, gRodon will omit temperature effects in the growth rate predictions.

<details>

<summary>What the input file looks like</summary>

| gene_location                                        | genome_name       | temperature |
|-------------------------------------------------|--------------------|-------------|
| known/RS_GCF_002749895.1/RS_GCF_002749895.1.ffn | RS_GCF_002749895.1 | 10          |
| known/RS_GCF_002849855.1/RS_GCF_002849855.1.ffn | RS_GCF_002849855.1 | 25          |
| known/RS_GCF_002906655.1/RS_GCF_002906655.1.ffn | RS_GCF_002906655.1 | 23          |
| known/RS_GCF_003026105.1/RS_GCF_003026105.1.ffn | RS_GCF_003026105.1 | 5          |
| known/RS_GCF_003026475.1/RS_GCF_003026475.1.ffn | RS_GCF_003026475.1 | 6          |
| known/RS_GCF_003026815.1/RS_GCF_003026815.1.ffn | RS_GCF_003026815.1 | 15         |
| known/RS_GCF_003144035.1/RS_GCF_003144035.1.ffn | RS_GCF_003144035.1 | 34          |
| known/RS_GCF_003544875.1/RS_GCF_003544875.1.ffn | RS_GCF_003544875.1 | 32          |
| known/RS_GCF_003716875.1/RS_GCF_003716875.1.ffn | RS_GCF_003716875.1 | 21          |
| known/RS_GCF_900130105.1/RS_GCF_900130105.1.ffn | RS_GCF_900130105.1 | 18          |
| known/RS_GCF_002749895.1/RS_GCF_002749895.1.ffn | 1                  | 10          |
| known/RS_GCF_002849855.1/RS_GCF_002849855.1.ffn | 2                  | 25          |
| known/RS_GCF_002906655.1/RS_GCF_002906655.1.ffn | 3                  | 23          |
| known/RS_GCF_003026105.1/RS_GCF_003026105.1.ffn | 4                  | 5          |
| known/RS_GCF_003026475.1/RS_GCF_003026475.1.ffn | 5                  | 6          |
| known/RS_GCF_003026815.1/RS_GCF_003026815.1.ffn | 6                  | 15         |
| known/RS_GCF_003144035.1/RS_GCF_003144035.1.ffn | 7                  | 34          |
| known/RS_GCF_003544875.1/RS_GCF_003544875.1.ffn | 8                  | 32          |
| known/RS_GCF_003716875.1/RS_GCF_003716875.1.ffn | 9                  | 21          |
| known/RS_GCF_900130105.1/RS_GCF_900130105.1.ffn | 10                 | 18          |

</details>


### `user_tree`: Identified/unidentified genomes in the GTDB 

The default value for `user_tree` is `NULL`, which assumes that the genomes are identified in the GTDB database.

-   **Identified genomes**: When users' genomes are identified in the GTDB and assigned accession numbers (e.g., `RS_GCF_002749895.1`), Phydon will search the GTDB for the representative genomes of the species that the users' genomes belong to. It will then leverage both phylogenetic information and gRodon to enhance the growth rate predictions.

-   **Unidentified genomes**: If the users' genomes are not identified in the GTDB, the user must provide a phylogenetic tree that integrates users' genomes in the phylogenetic tree from the GTDB. In this case, Phydon will compute the phylogenetic distance between the users' genomes and the genomes in the training data to estimate the maximum growth rates. Otherwise, Phydon will only use gRodon predictions.

### `tax`: Bacteria and Archaea 

By default, the `tax` argument is set to "bacteria". Users can change this to "archaea" if they wish to estimate the maximum growth rates of `archaeal` genomes.

Phydon supports both bacterial and archaeal genomes. Users must specify the taxon by setting the `tax` argument to either "bacteria" or "archaea" in order to generate valid growth rate estimates. If the taxon is not specified correctly, the predictions will return NA. The default is "bacteria".

### `regression_mode`: The functional types of the regression model

By default, regression_mode is set to "geometric_mean". The possible values for this argument are "geometric_mean" and "arithmetic_mean" that use models with a continuous P. Or "geometric_mean_binaryP" and "arithmetic_mean_binaryP" that use models with a binary P use models with a binary P.

Phydon offers two regression models for estimating the maximum growth rates of bacterial or archaeal genomes. The **arithmetic regression model with a continuous P** is based on the following formula (also see [Our paper](https://www.biorxiv.org/content/10.1101/2024.10.03.616540v1)):

```math
\tilde{y}_{Phydon} = P \times \tilde{y}_{gRodon} + (1-P) \times \tilde{y}_{phylo}
```

and the other is the **geometric regression model**

```math
\tilde{y}_{Phydon} = \tilde{y}_{gRodon}^P \times \tilde{y}_{phylo}^{(1-P)}
```

where $P$ is probability of the gRodon predictions outcompeting the Phylopred predictions. Users can choose between these models based on the nature of their data. The default is "geometric_mean".

We also include the model using a binary P denoted by "geometric_mean_binaryP" or "arithmetic_mean_binaryP", although we do not recommend it as instability occurs. 

```math
\tilde{y}_{Phydon} = \tilde{y}_{gRodon} \times I(P > 0.5) + \tilde{y}_{phylo} \times I(P <= 0.5)
```

### `gRodon_mode`: The gRodon mode

By default, `gRodon_mode` is set to "full". The possible values are "full" and "metagenome".

Phydon provides two modes for estimating the maximum growth rates of bacterial and archaeal genomes:

- Full mode:

This mode incorporates CUBHE, ConsistencyHE, and CPB in the regression model, allowing for a comprehensive estimation of growth rates.

- Metagenome mode:

This mode only uses CUBHE in the regression model to prevent unreliable estimations that may arise from metagenomic data, as the consistency statistic is not appropriate in this context. Users should be aware that the metagenome mode is expected to be less accurate than the full mode, and it should only be used when necessary.





# Output

The output of Phydon is a **data frame** containing the following columns:

-   `genome`: The accession number/names of the genome for which users aim to predict the growth rate.

-   `rep_genome`: The representative genome of the species to which the genome belongs, based on the GTDB database. If this value is NA, no matching representative genomes were found, and the genome will be treated as a unique species outside of the GTDB database. In this case, only gRodon predictions will be used.

-   `species`: The species name of the genome based on the GTDB database. If this value is `NA`, no matching species were found.

-   `neighbor_repgenome_train`: The accession number of the representative genome that is most phylogenetically close to the user's genome, based on the package's training data. If this value is `NA`, no matching representative genomes were found, and the phylogenetic distance will be set to infinity (`Inf`).

-   `phy_distance`: The phylogenetic distance of the user's genome to the database. If this value is `Inf`, no matching representative genomes were found.

-   `gRodonpred`:The gRodon predictions for the genome, based on the [gRodon2](https://github.com/jlw-ecoevo/gRodon2) package.

-   `phylopred`: The phylogenetic predictions for the genome, based on the [picante](https://www.rdocumentation.org/packages/picante/versions/1.8.2) package.

-   `combopred`: The combined predictions for the genome, integrating both gRodon and phylogenetic predictions. The regression model is trained based on the data in the package. See details in our [paper](https://www.biorxiv.org/content/10.1101/2024.10.03.616540v1).

Additionally, the output includes several useful metrics inherited from the gRodon package:


- `CUBHE`: The mean codon usage bias of each highly expressed gene relative to all other genes.

- `ConsistencyHE`: The mean codon usage bias of each highly expressed gene relative to all other highly expressed genes.

- `CPB`: The genome-wide codon pair bias.

- `FilteredSequences`: The number of sequences filtered due to length.

- `LowerCI` and `UpperCI`: The 95% confidence intervals for the gRodon estimation.



# Get A Phylogenetic Tree for Unidentified Genomes


> [!IMPORTANT]\
> When users' genomes are not identified in the GTDB database, they must provide a phylogenetic tree that includes both their genomes and the representative genomes from GTDB. This can be accomplished using [GTDB-Tk](https://ecogenomics.github.io/GTDBTk/index.html). To obtain the phylogenetic tree, run the [`classify_wf`](https://ecogenomics.github.io/GTDBTk/commands/classify_wf.html) command on your genomes as follows:

```         
gtdbtk classify_wf --skip_ani_screen --genome_dir /path/to/genomes --out_dir /output/
```

> The phylogenetic tree will be generated in the `/output/classify/` folder and is typically named `gtdbtk.bac120.classify.tree.1.tree`. Users should use this tree as the input for Phydon.


# Summary of Workflow

## Identified Genomes

1.  Use Prokka to annotate the genomes, obtaining the `.ffn` and `.gff` files. Optionally, generate the `genome1_CDS_names.txt` file (not necessary if you run Phydon on Linux or macOS).

2.  Structure the genomic data in the `genefiles/` directory and create the `data_info.csv` file as described above.

3.  Execute Phydon using the `data_info.csv` file.

## Unidentified Genomes

1.  Use GTDB-Tk to run the `classify_wf` command to create the `user_tree`.

2.  Use Prokka to annotate the genomes, obtaining the `.ffn` and `.gff` files. Optionally, generate the `genome1_CDS_names.txt` file (not necessary if you run Phydon on Linux or macOS).

3.  Structure the genomic data in the `genefiles/` directory and create the `data_info.csv` file as described above.

4.  Execute Phydon using the `data_info.csv` file and the `user_tree`.




# Processing A Large Number of Genomes in Parallel

In some cases, users may have a large number of genomes to estimate. To facilitate this process, we provide an example script that can be used to estimate multiple genomes in parallel. Below is the script:

```{r, eval = FALSE}
library(Phydon)
library(foreach)
library(doParallel)

### read the data
data_info <- read.csv(system.file("extdata", "known.csv", package = "Phydon"))
### To get the correct directory of the genomic data on the local computer
genomes <- data_info$genome_name
gene_location <- c()
for (i in 1:length(genomes)) {
  gene_location[i] <- system.file("extdata", paste0("known/",genomes[i],"/",genomes[i],".ffn"), package = "Phydon")
}
data_info$gene_location <- gene_location



num_cores <- 4
# parallel processing phydon to the data_info
cl <- makeCluster(num_cores)
registerDoParallel(cl)
phydon_res <- foreach(i = 1:nrow(data_info), .combine = rbind, .packages = "Phydon", .errorhandling = 'remove') %dopar% {
  Phydon::Phydon(data_info[i,])
}
stopCluster(cl)

phydon_res
# save
# write.csv(phydon_res, "phydon_prediction_results.csv", row.names=FALSE)
```

This script allows users to run Phydon for each genome in the specified directory in parallel, which can significantly speed up the estimation process.

# Example Script for Converting ASV Sequences to Accession Numbers in the GTDB

We recommend using the GTDB database for querying your ASV sequences. The following script can be used to download the GTDB database and convert ASV sequences to the accession numbers of the genomes in the GTDB database.


**Step 1: Download the GTDB Database and Create a BLAST Database**


```sh
#!/bin/bash -i 

### Create a conda environment with blast and download GTDB database for blast
#conda create -n blast-env -c bioconda blast

### Activate the environment and download the GTDB database
conda activate blast-env

#downloaded as follows into GTDB-blastdb-r220 directory:
#wget https://data.gtdb.ecogenomic.org/releases/release220/220.0/genomic_files_all/ssu_all_r220.fna.gz

# Create a BLAST database
makeblastdb -dbtype nucl -in GTDB-blastdb-r220/ssu_all_r220.fna 

```

This script downloads the GTDB database and creates a BLAST database.


**Step 2: Convert ASV Sequences to Accession Numbers**

Users can then use the following script to convert the ASV sequences to the accession numbers of the genomes in the GTDB database:

```sh
#!/bin/bash -i 
conda activate blast-env


mkdir -p BLAST-output-r220

#blast against GTDB r220
blastn -query your_sequences.fasta -db GTDB-blastdb-r220/ssu_all_r220.fna -outfmt 6 -perc_identity 95 -qcov_hsp_perc 100 > BLAST-output-r220/your_BLASTOUTPUT-95pcID-vs-GTDB-r220-allssu.tsv

### zip the output
#zip -m BLAST-output-r220/your_BLASTOUTPUT-95pcID-vs-GTDB-r220-allssu.zip BLAST-output-r220/your_BLASTOUTPUT-95pcID-vs-GTDB-r220-allssu.tsv

```

The output of this script will be a tab-delimited file containing the accession numbers of the genomes in the GTDB database along with several statistics. Users can then extract the accession numbers and fetch trait data from the trait database. 



# Citation

Liang Xu, Emily Zakem, JL Weissman, Improved maximum growth rate prediction from microbial genomes by integrating phylogenetic information. bioRxiv 2024.10.03.616540; doi: https://doi.org/10.1101/2024.10.03.616540
